src/validationinterface.cpp

```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
     boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
 
     CScheduler *scheduler = NULL;
+
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    bool fCallbacksRunning = false;
+
+    void MaybeScheduleProcessQueue() {
+        {
+            LOCK(cs_callbacksPending);
+            // Try to avoid scheduling too many copies here, but if we
+            // accidentally have two ProcessQueue's scheduled at once its
+            // not a big deal.
+            if (fCallbacksRunning) return;
+            if (callbacksPending.empty()) return;
+        }
+        scheduler->schedule(std::bind(&CMainSignalsInstance::ProcessQueue, this));
+    }
+
+    void ProcessQueue() {
+        std::function<void (void)> callback;
+        {
+            LOCK(cs_callbacksPending);
+            if (fCallbacksRunning) return;
+            if (callbacksPending.empty()) return;
+            fCallbacksRunning = true;
+
+            callback = callbacksPending.front();
+            callbacksPending.pop_front();
+        }
+
+        // RAII the setting of fCallbacksRunning and calling MaybeScheduleProcessQueue
```

Comment:
> Why not try{} catch{}?

My guess about this was that it allows the processqueue to take advantage of whatever error handling or reporting cscheduler provides, and to not have to repeat the finalization logic both inside and after the catch clause. Either approach seems fine to me, though.